File: gateway/src/services/supabaseStorage.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.SUPABASE_ANON_KEY;
const BUCKET_NAME = process.env.SUPABASE_BUCKET_NAME || 'crimeos-evidence';

let supabase = null;

if (SUPABASE_URL && SUPABASE_KEY) {
  try {
    supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
    console.log('[+] Supabase Storage client initialized successfully');
  } catch (err) {
    console.warn('[-] Failed to initialize Supabase Storage client:', err.message);
  }
} else {
  console.log('[i] Supabase Storage keys not provided. System will use local file storage/fallback.');
}

/**
 * Uploads a local file buffer or disk file to Supabase Storage bucket.
 * @param {string} localFilePath - Path to local file on disk
 * @param {string} storageDestinationPath - Storage path inside bucket (e.g. "complaints/CR-2026-9910.pdf")
 * @param {string} mimeType - File mime type
 * @returns {Promise<{ storageUrl: string, path: string, isCloud: boolean }>}
 */
const uploadFileToStorage = async (localFilePath, storageDestinationPath, mimeType = 'application/octet-stream') => {
  if (supabase && fs.existsSync(localFilePath)) {
    try {
      const fileBuffer = fs.readFileSync(localFilePath);
      
      // Ensure bucket exists or handle upload
      const { data, error } = await supabase.storage
        .from(BUCKET_NAME)
        .upload(storageDestinationPath, fileBuffer, {
          contentType: mimeType,
          upsert: true
        });

      if (error) {
        console.warn('[-] Supabase upload error:', error.message);
        throw error;
      }

      // Get public URL
      const { data: publicUrlData } = supabase.storage
        .from(BUCKET_NAME)
        .getPublicUrl(storageDestinationPath);

      console.log(`[+] File uploaded to Supabase Storage: ${publicUrlData.publicUrl}`);
      return {
        storageUrl: publicUrlData.publicUrl,
        path: data.path,
        isCloud: true
      };
    } catch (err) {
      console.warn('[-] Supabase upload fallback to local URL:', err.message);
    }
  }

  // Fallback to relative API endpoint URL
  const filename = path.basename(storageDestinationPath);
  return {
    storageUrl: `/api/requests/download/${filename}`,
    path: storageDestinationPath,
    isCloud: false
  };
};

/**
 * Generates a signed temporary download URL for a file in Supabase Storage.
 */
const getSignedDownloadUrl = async (storageDestinationPath, expiresInSeconds = 3600) => {
  if (supabase) {
    try {
      const { data, error } = await supabase.storage
        .from(BUCKET_NAME)
        .createSignedUrl(storageDestinationPath, expiresInSeconds);

      if (data && data.signedUrl) {
        return data.signedUrl;
      }
    } catch (err) {
      console.warn('[-] Supabase signed URL error:', err.message);
    }
  }
  const filename = path.basename(storageDestinationPath);
  return `/api/requests/download/${filename}`;
};

module.exports = {
  uploadFileToStorage,
  getSignedDownloadUrl,
  BUCKET_NAME
};

File: ai-service/app/services/supabase_storage.py
import os
import re
import hashlib
import requests
import mimetypes
from typing import Optional, Dict, Any, Union

SUPABASE_URL = os.environ.get(" ", "https://mimmgklddepndpgohcwo.supabase.co")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY") or os.environ.get("SUPABASE_ANON_KEY")
BUCKET_NAME = os.environ.get("SUPABASE_BUCKET_NAME", "crimeos-evidence")


def compute_sha256(data: Union[bytes, str]) -> str:
    """Computes SHA-256 hash of data for forensic tamper verification."""
    raw_bytes = data if isinstance(data, bytes) else data.encode("utf-8")
    return hashlib.sha256(raw_bytes).hexdigest()


def upload_to_supabase_storage(
    content_or_path: Union[bytes, str],
    destination_path: str,
    mime_type: Optional[str] = None
) -> Dict[str, Any]:
    """
    Uploads a file or binary payload to Supabase Storage with deterministic upsert.
    Guarantees no duplicate storage consumption (x-upsert: true).
    
    Returns:
        Dict with storage_url, path, sha256, and is_cloud flag.
    """
    # 1. Read file bytes
    if isinstance(content_or_path, str) and os.path.exists(content_or_path):
        with open(content_or_path, "rb") as f:
            file_bytes = f.read()
        if not mime_type:
            mime_type = mimetypes.guess_type(content_or_path)[0] or "application/octet-stream"
    elif isinstance(content_or_path, bytes):
        file_bytes = content_or_path
        if not mime_type:
            mime_type = "application/octet-stream"
    else:
        file_bytes = str(content_or_path).encode("utf-8")
        if not mime_type:
            mime_type = "text/plain; charset=utf-8"

    sha256_hash = compute_sha256(file_bytes)

    # Clean destination path (prevent spaces and special characters)
    clean_dest = re.sub(r'[^a-zA-Z0-9_./-]', '_', destination_path).lstrip('/')

    if SUPABASE_URL and SUPABASE_KEY:
        try:
            # Supabase Storage REST endpoint for object upload
            url = f"{SUPABASE_URL.rstrip('/')}/storage/v1/object/{BUCKET_NAME}/{clean_dest}"
            headers = {
                "Authorization": f"Bearer {SUPABASE_KEY}",
                "apikey": SUPABASE_KEY,
                "Content-Type": mime_type,
                "x-upsert": "true"  # Never creates duplicate files in Supabase
            }

            resp = requests.post(url, headers=headers, data=file_bytes, timeout=15)
            if resp.status_code in [200, 201]:
                public_url = f"{SUPABASE_URL.rstrip('/')}/storage/v1/object/public/{BUCKET_NAME}/{clean_dest}"
                return {
                    "storage_url": public_url,
                    "path": clean_dest,
                    "sha256": sha256_hash,
                    "is_cloud": True,
                    "status": "success"
                }
            else:
                # If bucket doesn't exist, log warning and fallback
                print(f"[-] Supabase upload warning ({resp.status_code}): {resp.text}")
        except Exception as e:
            print(f"[-] Supabase connection error: {e}")

    # Fallback to local / API relative URL
    return {
        "storage_url": f"/api/requests/download/{os.path.basename(clean_dest)}",
        "path": clean_dest,
        "sha256": sha256_hash,
        "is_cloud": False,
        "status": "local_fallback"
    }
